src/main/java/com/github/dikhan/pagerduty/client/events/HttpApiServiceImpl.java

```diff
@@ -43,42 +60,82 @@ private void initUnirestWithProxy(String proxyHost, Integer proxyPort) {
         try {
             HttpRequestWithBody request = Unirest.post(eventApi)
                     .header("Accept", "application/json");
             request.body(incident);
             HttpResponse<JsonNode> jsonResponse = request.asJson();
-            log.debug(IOUtils.toString(jsonResponse.getRawBody()));
-            switch(jsonResponse.getStatus()) {
+
+            if (log.isDebugEnabled()) {
+                log.debug(IOUtils.toString(jsonResponse.getRawBody()));
+                // A reset, so we can get the contents from the body that were dumped in the log before
+                jsonResponse.getRawBody().reset();
+            }
+
+            int responseStatus = jsonResponse.getStatus();
+            switch(responseStatus) {
                 case HttpStatus.SC_OK:
                 case HttpStatus.SC_CREATED:
                 case HttpStatus.SC_ACCEPTED:
                     return EventResult.successEvent(JsonUtils.getPropertyValue(jsonResponse, "status"), JsonUtils.getPropertyValue(jsonResponse, "message"), JsonUtils.getPropertyValue(jsonResponse, "dedup_key"));
                 case HttpStatus.SC_BAD_REQUEST:
                     return EventResult.errorEvent(JsonUtils.getPropertyValue(jsonResponse, "status"), JsonUtils.getPropertyValue(jsonResponse, "message"), JsonUtils.getArrayValue(jsonResponse, "errors"));
+                case RATE_LIMIT_STATUS_CODE:
+                case HttpStatus.SC_INTERNAL_SERVER_ERROR:
+                    if (doRetries) {
+                        return handleRetries(incident, retryCount, jsonResponse, responseStatus);
+                    } else {
+                        return EventResult.errorEvent(String.valueOf(responseStatus), "", IOUtils.toString(jsonResponse.getRawBody()));
+                    }
                 default:
-                    return EventResult.errorEvent(String.valueOf(jsonResponse.getStatus()), "", IOUtils.toString(jsonResponse.getRawBody()));
+                    return EventResult.errorEvent(String.valueOf(responseStatus), "", IOUtils.toString(jsonResponse.getRawBody()));
             }
         } catch (UnirestException | IOException e) {
             throw new NotifyEventException(e);
         }
     }
 
+    private EventResult handleRetries(Incident incident, int retryCount, HttpResponse<JsonNode> jsonResponse, int responseStatus) throws IOException, NotifyEventException {
+        long[] retryDelays = RETRY_WAIT_TIME_MILLISECONDS.get(responseStatus);
+
+        int maxRetries = retryDelays.length;
+        if (retryCount == maxRetries) {
+            log.debug("Received a {} response. Exhausted all the possibilities to retry.", responseStatus);
+            return EventResult.errorEvent(String.valueOf(responseStatus), "", IOUtils.toString(jsonResponse.getRawBody()));
+        }
+
+        log.debug("Received a {} response. Will retry again. ({}/{})", RATE_LIMIT_STATUS_CODE, retryCount, maxRetries);
```

Comment:
handleRetries is called from the notifyEvent function when the responseStatus is either RATE_LIMIT_STATUS_CODE or SC_INTERNAL_SERVER_ERROR. Hence should this logging entry output the responseStatus instead of hardcoding the code?

Reply:
Yes! It should definitively be like that.

Reply:
Yes! It should definitively be like that.